packages/common-sdk/src/web3/network/fetcher/simple-fetcher-impl.ts

```diff
@@ -0,0 +1,142 @@
+
+    const cached = this.cache.get(addressStr);
+    const maxAge = opts?.maxAge ?? this.retentionPolicy.get(parser) ?? Number.POSITIVE_INFINITY;
+    const elapsed = !!cached ? now - (cached?.fetchedAt ?? 0) : Number.NEGATIVE_INFINITY;
+    const expired = elapsed > maxAge;
+
+    if (!!cached && !expired) {
+      return cached.value as U | null;
+    }
+
+    try {
+      const accountInfo = await this.connection.getAccountInfo(addressKey);
+      const value = parser.parse(addressKey, accountInfo);
+      this.cache.set(addressStr, { parser, value, fetchedAt: now });
+      return value;
+    } catch (e) {
+      this.cache.set(addressStr, { parser, value: null, fetchedAt: now });
+      return null;
+    }
+  }
+
+  async getAccounts<U extends T>(
+    addresses: Address[],
+    parser: ParsableEntity<U>,
+    opts?: SimpleAccountFetchOptions | undefined,
+    now: number = Date.now()
+  ): Promise<ReadonlyMap<string, U | null>> {
+    const addressStrs = AddressUtil.toStrings(addresses);
+    await this.populateCache(addressStrs, parser, opts, now);
+
+    // Build a map of the results, insert by the order of the addresses parameter
+    const result = new Map<string, U | null>();
+    addressStrs.forEach((addressStr) => {
+      const cached = this.cache.get(addressStr);
+      const value = cached?.value as U | null;
+      result.set(addressStr, value);
+    });
+
+    // invariant(result.size === addresses.length, "not enough results fetched");
+    return result;
+  }
+
+  async getAccountsAsArray<U extends T>(
+    addresses: Address[],
+    parser: ParsableEntity<U>,
+    opts?: FetchOptions | undefined,
+    now: number = Date.now()
+  ): Promise<ReadonlyArray<U | null>> {
+    const addressStrs = AddressUtil.toStrings(addresses);
+    await this.populateCache(addressStrs, parser, opts, now);
+
+    // Rebuild an array containing the results, insert by the order of the addresses parameter
+    const result = new Array<U | null>();
+    addressStrs.forEach((addressStr) => {
+      const cached = this.cache.get(addressStr);
+      const value = cached?.value as U | null;
+      result.push(value);
+    });
+
+    return result;
+  }
+
+  async refreshAll(now: number = Date.now()) {
+    const addresses = Array.from(this.cache.keys());
+    const fetchedAccountsMap = await getMultipleAccountsInMap(this.connection, addresses);
+
+    for (const [key, cachedContent] of this.cache.entries()) {
+      const parser = cachedContent.parser;
+      const fetchedEntry = fetchedAccountsMap.get(key);
+      const value = parser.parse(AddressUtil.toPubKey(key), fetchedEntry);
+      this.cache.set(key, { parser, value, fetchedAt: now });
+    }
+  }
+  private async populateCache<U extends T>(
+    addresses: Address[],
+    parser: ParsableEntity<U>,
+    opts?: SimpleAccountFetchOptions | undefined,
+    now: number = Date.now()
+  ) {
+    const addressStrs = AddressUtil.toStrings(addresses);
```

Comment:
nit: since this is in a private method, we can probably use `PublicKey[]` as the type for `addresses` and skip this transformation and ensure we do the transform at the caller

Reply:
The transformation is conditional and the transformation is pretty slow. If we force the transformation back to PublicKey, then the getAccounts methods will have to perform the transformation twice, which can be quite costly

Reply:
Ah yes gotcha, makes sense since it's just transforming everything to strings